docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,555 @@
+another resource. This satisfies the use case of referencing 'ID(s)/Name(s)/ARN(s)'
+from other resources
+
+### Future Improvements
+* Add support for custom hooks while dereferencing resources
+* Add support for `selector` for referencing the objects
+* Add support for references in custom fields that are not present in
+* Add support for referencing resource from different api-groups
+* Add support for not deleting a resource if it is referenced in another resource
+* Add support for finite retries and terminal conditions when resolving references
+
+## Preferred Approach
+
+### High Level Overview
+
+* Introduce new construct in generator.yaml which will indicate to code-generator
+if a new spec field with type `AWSResourceReferenceWrapper` or
+`AWSResourcesReferenceWrapper` should be added to the custom resource definition.
+
+* This new spec field will work as additional input for an existing spec field.
+Addition of the extra spec field keeps the backward compatibility with existing
+custom resource definitions.
+
+* Validation will be present that only one of existing spec field OR new reference
+field is present in custom resource, not both.
+
+* If the existing spec field was marked as required, validation will be added that
+at least one of existing spec field or new reference field is present.
+
+* If `AWSResourceReferenceWrapper` type field is present in the desired state,
+first ACK controller will query for the referenced resource. If the referenced resource
+has `ACK.ResourceSynced` condition status `True` & specified field is found
+in the referenced resource, the reconciliation loop will progress. Otherwise
+the request will get requeued for future processing.
+
+* Watchers for the referenced resource type will not be implemented in the initial
+implementation. The initial implementation will mostly be used for finding
+identifier of referenced resource and any update in referenced resource should not
+change it's identifier. As mentioned earlier, deletion will also be handled in
+future.
+
+* The code to determine when the referenced resource is synced will be generated,
+and additional hooks will be added for customization. Ex: Do not just check presence
+of a field but also check if the Status is available etc...
+
+### Low Level Implementation
+
+#### generator.yaml Changes
+
+```go
+type FieldConfig struct {
+    ...
+    References *ReferencesConfig `json:"references,omitempty"`
+    ...
+}
+
+// ReferencesConfig contains the instructions for how to add the referenced resource
+// configuration for a field.
+type ReferencesConfig struct {
+	// Resource mentions the K8s resource which is read to resolve the
+	// reference
+	Resource string `json:"resource"`
+	// Path refers to the the path of field which should be copied
+	// to resolve the reference
+	Path string `json:"path"`
+}
+```
+
+`APIGatewayv2 generator.yaml` Example:
+```yaml
+resources:
+  Integration:
+    fields:
+      ApiId:
```

Comment:
Should this be `APIID`? The name of the spec field?

Reply:
The code inside `crd.go` works with "Original" field names not "Camel". Hence generator.yaml has field names as their original name in aws-sdk-go. To change this, We'll need to update logic inside `crd.go`

```go
func (r *CRD) AddSpecField(
	memberNames names.Names,
	shapeRef *awssdkmodel.ShapeRef,
) {
	fPath := memberNames.Camel
	fConfigs := r.cfg.ResourceFields(r.Names.Original)
	fConfig := fConfigs[memberNames.Original]
	f := NewField(r, fPath, memberNames, shapeRef, fConfig)
	if fConfig != nil && fConfig.Print != nil {
		r.addSpecPrintableColumn(f)
	}
	r.SpecFields[memberNames.Original] = f
	r.Fields[fPath] = f
}```